internal/metadata/metadata.go

```diff
@@ -82,30 +82,14 @@ func Set(addr resolver.Address, md metadata.MD) resolver.Address {
 // - header names must contain one or more characters from this set [0-9 a-z _ - .].
 // - if the header-name ends with a "-bin" suffix, no validation of the header value is performed.
 // - otherwise, the header value must contain one or more characters from the set [%x20-%x7E].
-func Validate(md metadata.MD) error {
+func Validate(md metadata.MD) (err error) {
 	for k, vals := range md {
-		// pseudo-header will be ignored
-		if k[0] == ':' {
-			continue
-		}
-		// check key, for i that saving a conversion if not using for range
-		for i := 0; i < len(k); i++ {
-			r := k[i]
-			if !(r >= 'a' && r <= 'z') && !(r >= '0' && r <= '9') && r != '.' && r != '-' && r != '_' {
-				return fmt.Errorf("header key %q contains illegal characters not in [0-9a-z-_.]", k)
-			}
-		}
-		if strings.HasSuffix(k, "-bin") {
-			continue
-		}
-		// check value
-		for _, val := range vals {
-			if hasNotPrintable(val) {
-				return fmt.Errorf("header key %q contains value with non-printable ASCII characters", k)
-			}
+		err = ValidatePair(k, vals...)
+		if err != nil {
+			break
 		}
 	}
-	return nil
+	return
```

Comment:
Please don't use implicit returns.  `return err` instead.

Also, optionally can be rewritten as:

```go
for k, vals := range md {
	if err := ValidatePair(k, vals...); err != nil {
		return err
	}
}
return nil
```